torch/_functorch/aot_autograd.py

```diff
@@ -138,86 +138,365 @@ def posthook(grad_input, grad_output):
+# # The info on which inputs are mutated is also tracked *before* synthetic base creation.
+# def compiled_wrapper(x, x_view):
+#     base = merge_view_inputs(x, x_view)
+#     x_updated, out = autograd.Function.apply(base)
+#     # x and x_view are aliased in eager mode, so this mutation to x will automatically affect x_view.
+#     x.copy_(x_updated)
+#     return out
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+# This class stores info about every "output:" where outputs are defined as either:
+# - outputs from the original user's forward()
+# - graph intermediates that we manually return in the compiled forward (but don't return to the user)
 @dataclass(frozen=True)
 class OutputAliasInfo:
     # Tells us if this output is:
     # (1) a regular (non-aliased) output
     # (2) an alias of a forward input
-    # (2) an alias of an intermediate (aka an alias of an output of the inner traced forward)
+    # (3) an alias of an intermediate (aka an alias of an output of the inner traced forward)
     output_type: OutputType
     # If (1) above, then
+    # - base_idx is None
+    # If (2) above, then
     # - Tells us that the base of this alias is user_fwd_input[base_idx]
     #   (This is an index into the inputs *before* we make synthetic bases)
-    # If (2) above, then
+    # If (3) above, then
     # - Tells us that the base of this alias is traced_fwd_outputs[base_idx]
     #   here, this refers to the index of the *direct* traced
-    base_idx: int
-    # sizes, strides and storage offset of the aliased output are all returned as actual (sym)ints
-    # in the compiled forward. These indices tell us where in the forward outputs to grab them.
-    sizes_idx: Optional[int]
-    strides_idx: Optional[int]
-    storage_offset_idx: Optional[int]
-    # We store the actual output alias that we traced in the forward (should be a fake tensor)
-    # to grab any other non-symbolic properties on the output alias, like requires_grad.
-    # It's optional here, for cases where the user directly returns an input as an output.
-    # If output_type == non_alias, then these fields are also always None.
-    tensor_meta: Optional[Tensor]
-
-# This class tells us about how to perform a metadata mutation on forward inputs.
-# it only applies to forward inputs that experience metadata-only mutations
+    base_idx: Optional[int]
+    # Only important when output_type == OutputType.alias_of_input
+    # Tells the runtime code to return args[base_idx] *exactly*, and not generate a view off of it.
+    is_input_tensor: bool
+    # Only important when output_type == OutputType.alias_of_intermediate
+    # Tells the code that traces the joint that we should append output._base as an output
+    # in the forward graph.
+    add_base_to_outputs: bool
```

Comment:
The fact that these booleans are only valid on specific settings of `output_type` suggests that maybe they shouldn't be booleans, and you just need more OutputTypes?

Reply:
that's fair, this is the "my current representation allows for invalid states" thing. I can add two new enum values like:
```
OutputType.is_input
# This one's pretty weird. The idea is that for every output that aliases an intermediate,
# we need the intermediate to be a graph output.
# We don't want to redundantly return a graph intermediate multiple times as a graph output though,
# so we only do it for some of them.
OutputType.alias_of_intermediate_where_base_must_be_returned
```

Originally, I felt that the other two values for these bools didn't really make sense for an output type, e.g. `OutputType.
 I tried to set things up so that the `OutputType` enum logically corresponding to

Reply:
looks like comment is truncated

Reply:
but yes, i'm a big fan of make illegal states unrepresentable